### PR TITLE
Human-friendly output when inspecting ActiveModel with attributes

### PR DESCRIPTION
**Pull request by Drenmi**
### Summary

**Background:**

We make liberal use of `ActiveModel` instances with the attributes API to model records originating outside our application. This works great, but the output when printing one of these records is really quite unusable. This PR implements human-readable output for classes' and instances' `#inspect` methods, mirroring that of "proper" ActiveRecord objects.

**Before:**

```ruby
#<ModelForAttributesTest:0x00007ffd77a8dc08
 @attributes=
  #<ActiveModel::AttributeSet:0x00007ffd77a8d988
   @attributes=
    {"integer_field"=>
      #<ActiveModel::Attribute::FromUser:0x00007ffd77a8d8e8
       @name="integer_field",
       @original_attribute=
        #<ActiveModel::Attribute::WithCastValue:0x00007ffd77a8db40
         @name="integer_field",
         @original_attribute=nil,
         @type=#<ActiveModel::Type::Integer:0x00007ffd756be778 @limit=nil, @precision=nil, @range=-2147483648...2147483648, @scale=nil>,
         @value_before_type_cast=nil>,
       @type=#<ActiveModel::Type::Integer:0x00007ffd756be778 @limit=nil, @precision=nil, @range=-2147483648...2147483648, @scale=nil>,
       @value_before_type_cast=42>,
     "string_field"=>
      #<ActiveModel::Attribute::FromUser:0x00007ffd77a8d870
       @name="string_field",
       @original_attribute=
        #<ActiveModel::Attribute::WithCastValue:0x00007ffd77a8dac8
         @name="string_field",
         @original_attribute=nil,
         @type=#<ActiveModel::Type::String:0x00007ffd756bc900 @false="f", @limit=nil, @precision=nil, @scale=nil, @true="t">,
         @value_before_type_cast=nil>,
       @type=#<ActiveModel::Type::String:0x00007ffd756bc900 @false="f", @limit=nil, @precision=nil, @scale=nil, @true="t">,
       @value_before_type_cast="Rails FTW">,
     "decimal_field"=>
      #<ActiveModel::Attribute::FromUser:0x00007ffd77a8d7d0
       @name="decimal_field",
       @original_attribute=
        #<ActiveModel::Attribute::WithCastValue:0x00007ffd77a8da78
         @name="decimal_field",
         @original_attribute=nil,
         @type=#<ActiveModel::Type::Decimal:0x00007ffd76fa3040 @limit=nil, @precision=nil, @scale=nil>,
         @value_before_type_cast=nil>,
       @type=#<ActiveModel::Type::Decimal:0x00007ffd76fa3040 @limit=nil, @precision=nil, @scale=nil>,
       @value_before_type_cast=12.3>,
     "string_with_default"=>
      #<ActiveModel::Attribute::UserProvidedDefault:0x00007ffd77a8da50
       @name="string_with_default",
       @original_attribute=
        #<ActiveModel::Attribute::UserProvidedDefault:0x00007ffd76fa0ed0
         @name="string_with_default",
         @original_attribute=nil,
         @type=#<ActiveModel::Type::String:0x00007ffd7882c1c0 @false="f", @limit=nil, @precision=nil, @scale=nil, @true="t">,
         @user_provided_value="default string",
         @value_before_type_cast="default string">,
       @type=#<ActiveModel::Type::String:0x00007ffd76fa1150 @false="f", @limit=nil, @precision=nil, @scale=nil, @true="t">,
       @user_provided_value="default string",
       @value_before_type_cast="default string">,
     "date_field"=>
      #<ActiveModel::Attribute::UserProvidedDefault:0x00007ffd77a8da28
       @name="date_field",
       @original_attribute=
        #<ActiveModel::Attribute::UserProvidedDefault:0x00007ffd76fabb00
         @name="date_field",
         @original_attribute=nil,
         @type=#<ActiveModel::Type::Date:0x00007ffd6f50e640 @limit=nil, @precision=nil, @scale=nil>,
         @user_provided_value=#<Proc:0x00007ffd6f50f1f8 (irb):8 (lambda)>,
         @value_before_type_cast=#<Proc:0x00007ffd6f50f1f8 (irb):8 (lambda)>>,
       @type=#<ActiveModel::Type::Date:0x00007ffd76fabda8 @limit=nil, @precision=nil, @scale=nil>,
       @user_provided_value=#<Proc:0x00007ffd76fa0980 (irb):25 (lambda)>,
       @value_before_type_cast=#<Proc:0x00007ffd76fa0980 (irb):25 (lambda)>>,
     "boolean_field"=>
      #<ActiveModel::Attribute::FromUser:0x00007ffd77a8d730
       @name="boolean_field",
       @original_attribute=
        #<ActiveModel::Attribute::WithCastValue:0x00007ffd77a8da00
         @name="boolean_field",
         @original_attribute=nil,
         @type=#<ActiveModel::Type::Boolean:0x00007ffd76faa6d8 @limit=nil, @precision=nil, @scale=nil>,
         @value_before_type_cast=nil>,
       @type=#<ActiveModel::Type::Boolean:0x00007ffd76faa6d8 @limit=nil, @precision=nil, @scale=nil>,
       @value_before_type_cast=true>,
     "value_field"=>
      #<ActiveModel::Attribute::FromUser:0x00007ffd77a8d618
       @name="value_field",
       @original_attribute=
        #<ActiveModel::Attribute::WithCastValue:0x00007ffd77a8d9b0
         @name="value_field",
         @original_attribute=nil,
         @type=#<ActiveModel::Type::Value:0x00007ffd76fa9c38 @limit=nil, @precision=nil, @scale=nil>,
         @value_before_type_cast=nil>,
       @type=#<ActiveModel::Type::Value:0x00007ffd76fa9c38 @limit=nil, @precision=nil, @scale=nil>,
       @value_before_type_cast="Yay">}>>
```

**After:**

```ruby
#<ModelForAttributesTest
 integer_field: 42,
 string_field: "Rails FTW",
 decimal_field: 0.123e2,
 string_with_default: "default string",
 date_field: "2016-01-01",
 boolean_field: true,
 value_field: "Yay">
```

### Other Information

**Limitations:**

This PR does not add attribute filtering. I originally intended to include that, but filter configuration lives in a very far away place in the inheritance chain, inside `ActiveRecord::Core`, and so requires either quite a bit of rejigging, or a separate filter configuration for `ActiveModel`.

_Note that this PR does not change anything about the output except the formatting. The existing implementation also does not filter any attributes._

For the same reason, there is no attempt to share the formatting functionality at the moment.

I am willing to work on this next with some direction.

### CODE REVIEW

File: activemodel/lib/active_model/attributes.rb [hunk 1]
```diff
@@ -42,6 +42,23 @@ def attribute_names
...
+      #   end
+      #
+      #  Person.inspect
+      #  # => Person(name: string, age: integer)
+      def inspect
+        attr_types = attribute_types.map { |name, type| "#{name}: #{type.type || 'value'}" }
+
+        "#{super}(#{attr_types.join(', ')})"
+      end
+
```
Comment (p8) [inline activemodel/lib/active_model/attributes.rb:61]: Same thing for this method.

File: activemodel/lib/active_model/attributes.rb [hunk 2]
```diff
@@ -117,6 +134,26 @@ def attribute_names
...
+    #  person.inspect
+    #  # => #<Person name: "David", age: 41>
+    def inspect
+      inspection = attributes.map do |name, value|
+        "#{name}: #{format_for_inspect(name, value)}"
+      end.join(", ")
+
+      "#<#{self.class} #{inspection}>"
+    end
+
```
Comment (p8) [inline activemodel/lib/active_model/attributes.rb:156]: This method seems like the only one that really changed. Not sure if it’s better have seperate implementations or if ActiveRecord could call super, in some cases, to call this method.

**Comment by Drenmi** [comment]
Hello, @p8! 👋 

Thanks for the review.

Yeah. So the one barrier to unification here is that ActiveRecord's inspect does filtering, and the filter configuration lives over in `ActiveRecord::Core`, which is not part of the inheritance chain for ActiveModel. I have been very careful not to make ActiveModel parts depend on ActiveRecord parts in any way.

So if we want to move the methods as part of this PR, I think we have two options:

1. Change the method in ActiveRecord's chain to call `#super` and do filtering as an additional step.
2. Do a little bit of an `if defined?` dance to check for the presence of filter methods.

The third option is to determine if and how filters should work together with ActiveModel and rework the filter part as well.

**Comment by p8** [comment]
@Drenmi thanks for looking into it.
Maybe it's better to keep the `inspect` implementations independent for now?

Is it possible to remove the `attribute_for_inspect` method from ActiveRecord?
https://github.com/rails/rails/blob/4eeca2af40c0853a9909414d6db3c44780d021f7/activerecord/lib/active_record/attribute_methods.rb#L268-L288

And remove the `format_for_inspect` method from ActiveRecord?
https://github.com/rails/rails/blob/4eeca2af40c0853a9909414d6db3c44780d021f7/activerecord/lib/active_record/attribute_methods.rb#L407-L421

**Comment by Drenmi** [comment]
> Is it possible to remove the `#attribute_for_inspect` method from ActiveRecord?

This can safely be removed if we're okay with the additional indirection.

> And remove the `#format_for_inspect` method from ActiveRecord?

This is where things get hairy. This is where the ActiveRecord version hooks into the parameter filter, which belongs to a separate inheritance chain away from ActiveModel. The relevant line is here: https://github.com/rails/rails/blob/4eeca2af40c0853a9909414d6db3c44780d021f7/activerecord/lib/active_record/attribute_methods.rb#L419

So this method can not be removed. There are two options I can think of after ruling out the ones that result in ActiveModel depending on ActiveRecord.

---

Either we call `#super`, like this:

```ruby
def format_for_inspect(name, value)
  if value.nil?
    value.inspect
  else
    inspection_filter.filter_param(name, super)
  end
end
```

Not so nice, due to the fact that the filtering only happens in one of the branches.

---

Or we extend the parameter filtering to ActiveModel as well, so the filter configuration lives in the ActiveModel part of the inheritance chain, and is applied the same for both ActiveRecord and ActiveModel.

I think this option probably warrants its own PR.

**Comment by p8** [comment]
@Drenmi Thanks for diving into this.
I think the `InspectionFilter` would be a nice addition for `ActiveModel`.
But I agree that is a lot bigger change that might be better in a separate PR.
So I think the first option would be best for now, as the current `format_for_inspect` has some pretty complex logic that we don't want to duplicate?

**Comment by Drenmi** [comment]
Great! I'll look at making the changes to this PR and then get cracking at the bigger one. 🙏

File: activemodel/lib/active_model/attribute_methods.rb [hunk 3]
```diff
@@ -547,6 +549,28 @@ def respond_to?(method, include_private_methods = false)
...
+    #
+    #   person.attribute_for_inspect(:name)
+    #   # => "\"David Heinemeier Hansson David Heinemeier Hansson ...\""
+    #
+    #   person.attribute_for_inspect(:created_at)
+    #   # => "\"2012-10-22 00:15:07.000000000 +0000\""
+    #
+    #   person.attribute_for_inspect(:tag_ids)
+    #   # => "[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]"
+    def attribute_for_inspect(attr_name)
```
Comment (rafaelfranca) [inline activemodel/lib/active_model/attribute_methods.rb:567]: Why do we have this method here? It is not used anywhere.

File: activemodel/lib/active_model/attribute_methods.rb [hunk 4]
```diff
@@ -567,6 +591,20 @@ def _read_attribute(attr)
         __send__(attr)
       end
 
+      def format_for_inspect(name, value)
```
Comment (rafaelfranca) [inline activemodel/lib/active_model/attribute_methods.rb:594]: Why is this method in this module and not in `Attribute`?

**Pull request closed without merging**